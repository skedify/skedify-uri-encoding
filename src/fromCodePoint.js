export default function fromCodePoint(...args) {
  const { length } = args;
  const MAX_SIZE = 0x4000;
  const codeUnits = [];
  let highSurrogate;
  let lowSurrogate;
  let index = -1;

  if (!length) {
    return '';
  }

  let result = '';

  while (++index < length) {
    let codePoint = Number(args[index]);

    if (
      !isFinite(codePoint) || // `NaN`, `+Infinity`, or `-Infinity`
      codePoint < 0 || // not a valid Unicode code point
      codePoint > 0x10ffff || // not a valid Unicode code point
      Math.floor(codePoint) !== codePoint // not an integer
    ) {
      throw new RangeError(`Invalid code point: ${codePoint}`);
    }

    if (codePoint <= 0xffff) {
      // BMP code point
      codeUnits.push(codePoint);
    } else {
      // Astral code point; split in surrogate halves
      // http://mathiasbynens.be/notes/javascript-encoding#surrogate-formulae
      codePoint -= 0x10000;
      highSurrogate = (codePoint >> 10) + 0xd800;
      lowSurrogate = (codePoint % 0x400) + 0xdc00;
      codeUnits.push(highSurrogate, lowSurrogate);
    }

    if (index + 1 === length || codeUnits.length > MAX_SIZE) {
      result += String.fromCharCode.apply(null, codeUnits);
      codeUnits.length = 0;
    }
  }

  return result;
}
