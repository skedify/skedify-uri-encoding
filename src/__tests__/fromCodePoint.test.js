import fromCodePoint from '../fromCodePoint'

const testCases = [
  { expect: fromCodePoint(''), toBe: '\0' },
  { expect: fromCodePoint(), toBe: '' },
  { expect: fromCodePoint(-0), toBe: '\0' },
  { expect: fromCodePoint(0), toBe: '\0' },
  { expect: fromCodePoint(0x1d306), toBe: '\uD834\uDF06' },
  {
    expect: fromCodePoint(0x1d306, 0x61, 0x1d307),
    toBe: '\uD834\uDF06a\uD834\uDF07',
  },
  { expect: fromCodePoint(0x61, 0x62, 0x1d307), toBe: 'ab\uD834\uDF07' },
  { expect: fromCodePoint(false), toBe: '\0' },
  { expect: fromCodePoint(null), toBe: '\0' },
]

const testCasesThrowsError = [
  { value: '_' },
  { value: '+Infinity' },
  { value: '-Infinity' },
  { value: -1 },
  { value: 0x10ffff + 1 },
  { value: 3.14 },
  { value: 3e-2 },
  { value: -Infinity },
  { value: Number(Infinity) },
  { value: NaN },
  { value: undefined },
  { value: {} },
  { value: /./ },
]

describe('fromCodePoint', () => {
  testCases.forEach(testCase => {
    test(`expect ${testCase.expect} toBe ${testCase.toBe}`, () => {
      expect(testCase.expect).toBe(testCase.toBe)
    })
  })

  testCasesThrowsError.forEach(testCase => {
    test(`expect ${testCase.value} to throw a RangeError`, () => {
      /* eslint-disable-next-line max-nested-callbacks */
      expect(() => fromCodePoint(testCase.value)).toThrowErrorMatchingSnapshot()
    })
  })
})
