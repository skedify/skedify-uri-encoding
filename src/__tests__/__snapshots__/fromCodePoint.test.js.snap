// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromCodePoint expect +Infinity to throw a RangeError 1`] = `"Invalid code point: Infinity"`;

exports[`fromCodePoint expect /./ to throw a RangeError 1`] = `"Invalid code point: NaN"`;

exports[`fromCodePoint expect [object Object] to throw a RangeError 1`] = `"Invalid code point: NaN"`;

exports[`fromCodePoint expect _ to throw a RangeError 1`] = `"Invalid code point: NaN"`;

exports[`fromCodePoint expect -1 to throw a RangeError 1`] = `"Invalid code point: -1"`;

exports[`fromCodePoint expect -Infinity to throw a RangeError 1`] = `"Invalid code point: -Infinity"`;

exports[`fromCodePoint expect -Infinity to throw a RangeError 2`] = `"Invalid code point: -Infinity"`;

exports[`fromCodePoint expect 0.03 to throw a RangeError 1`] = `"Invalid code point: 0.03"`;

exports[`fromCodePoint expect 3.14 to throw a RangeError 1`] = `"Invalid code point: 3.14"`;

exports[`fromCodePoint expect 1114112 to throw a RangeError 1`] = `"Invalid code point: 1114112"`;

exports[`fromCodePoint expect Infinity to throw a RangeError 1`] = `"Invalid code point: Infinity"`;

exports[`fromCodePoint expect NaN to throw a RangeError 1`] = `"Invalid code point: NaN"`;

exports[`fromCodePoint expect undefined to throw a RangeError 1`] = `"Invalid code point: NaN"`;
